internal/bufferedwrites/upload_handler.go

```diff
@@ -154,3 +155,39 @@ func (uh *UploadHandler) SignalUploadFailure() chan error {
 func (uh *UploadHandler) AwaitBlocksUpload() {
 	uh.wg.Wait()
 }
+
+func (uh *UploadHandler) Destroy() {
+	// Waiting for upload routine to move all blocks to freeChannel.
```

Comment:
I think we need something like this here:

```
	// Attempt to close the upload channel in case it is not closed.
	func() {
		defer func() {
			_ = recover()
		}()
		close(uh.uploadCh)
	}()
```

if we won't close the upload channel, then the cleanup for loop will never exit.
```
for currBlock := range uh.uploadCh {
			uh.freeBlocksCh <- currBlock
			// Marking as wg.Done to ensure any waiters are unblocked.
			uh.wg.Done()
		}
```

Reply:
Changed the logic to handle it via select statement.